File: backend/ProjectProgressManagementSystem/Services/Implementations/ProjectService.cs
using Microsoft.EntityFrameworkCore;
using ProjectProgressManagementSystem.DataAccess;
using ProjectProgressManagementSystem.Exceptions;
using ProjectProgressManagementSystem.Extensions;
using ProjectProgressManagementSystem.Maps.ModelMappers;
using ProjectProgressManagementSystem.Models;
using ProjectProgressManagementSystem.Services.Interfaces;
using ProjectProgressManagementSystem.ViewModels;
using ProjectProgressManagementSystem.ViewModels.Project;
using System.Data;

namespace ProjectProgressManagementSystem.Services.Implementations
{
    public class ProjectService : IProjectService
    {
        private readonly ApplicationDbContext _context;
        public ProjectService(ApplicationDbContext context)
        {
            _context = context;
        }

        // get all projects
        public async Task<IEnumerable<ProjectViewModel>> GetAllAsync(Department? department, Region? region)
        {
            return await _context.Projects
                .Where(project => department == null || (department.Value & project.Department) > 0)
                .Where(project => region == null || (region.Value & project.Region) > 0)
                .Where(project => project.DateDeleted == null || project.DateDeleted >= DateTime.Now)
                .Select(project => new ProjectViewModel
                {
                    Id = project.Id,
                    Number = project.Number,
                    Title = project.Title,
                    Business = project.Business,
                    Department = project.Department,
                    Region = project.Region,
                    Description = project.Description,
                }).ToListAsync();
        }

        // get all projects in month, year
        public async Task<IEnumerable<ProjectViewModel>> GetAllAsync(int year, int month, Department? department, Region? region)
        {
            var firstDayOfNextMonth = new DateTime(year, month, 1).AddMonths(1);
            return await _context.Projects
                .Where(project => department == null || (department.Value & project.Department) > 0)
                .Where(project => region == null || (region.Value & project.Region) > 0)
                .Where(project => project.DateDeleted == null || project.DateDeleted >= firstDayOfNextMonth)
                .Where(project => project.DateCreated < firstDayOfNextMonth)
                .Select(project => new ProjectViewModel
                {
                    Id = project.Id,
                    Number = project.Number,
                    Title = project.Title,
                    Business = project.Business,
                    Department = project.Department,
                    Region = project.Region,
                    Description = project.Description,
                }).ToListAsync();
        }

        // get project by id
        public async Task<ProjectViewModel> GetByIdAsync(int id)
        {
            var project = await FromId(id);
            return ViewModelMapper.ToViewModel(project);
        }

        // create a new project
        public async Task<ProjectViewModel> CreateAsync(ProjectCreateViewModel project)
        {
            var existingProject = await _context.Projects.AnyAsync(p => p.DateDeleted == null && (p.Number == project.Number || p.Title == project.Title));
            if (existingProject)
            {
                throw new DuplicateEntityException($"Project already exists in {project.Department} department.");
            }
            var projectEntity = EntityModelMapper.ToEntity(project);
            await _context.Projects.AddAsync(projectEntity);
            await _context.SaveChangesAsync();
            return ViewModelMapper.ToViewModel(projectEntity);
        }

        // import projects from excel
        public async Task<MessageViewModel> ImportFromExcelAsync(IFormFile excelFile)
        {
            try
            {
                // Validate the file
                if (excelFile == null || excelFile.Length == 0)
                {
                    throw new ArgumentException("Excel file is required.");
                }

                var dataTable = ExcelReader.ReadExcelFile(excelFile);
                var projects = new List<Project>();

                foreach (DataRow row in dataTable.Rows)
                {
                    var project = EntityModelMapper.ToProjectEntity(row);
                    projects.Add(project);
                }
                // Add users to the database
                await _context.Projects.AddRangeAsync(projects);
                await _context.SaveChangesAsync();

                return new MessageViewModel
                {
                    Message = "Projects imported successfully."
                };
            }
            catch (Exception ex)
            {
                throw new Exception($"Error importing projects from Excel: {ex.Message}");
            }
        }

        // update an existing project
        public async Task<ProjectViewModel> UpdateAsync(int id, ProjectUpdateViewModel project)
        {
            var projectEntity = await FromId(id);
            if (projectEntity.DateDeleted != null)
            {
                throw new OperationNotSupportedException($"Project with id {id} is scheduled to delete on {projectEntity.DateDeleted}.");
            }
            if (project.Number == null && project.Title == null)
            {
                projectEntity.Department = project.Department ?? projectEntity.Department;
                projectEntity.Description = project.Description ?? projectEntity.Description;
                await _context.SaveChangesAsync();
                return ViewModelMapper.ToViewModel(projectEntity);
            }
            var today = DateTime.Now;
            var lastDateOfPreviousMonth = new DateTime(today.Year, today.Month-1, DateTime.DaysInMonth(today.Year, today.Month-1));
            projectEntity.DateModified = DateTime.Now.Date;
            projectEntity.DateDeleted = lastDateOfPreviousMonth;
            var newProjectEntity = EntityModelMapper.ToEntity(projectEntity, project);
            await _context.Projects.AddAsync(newProjectEntity);
            await _context.SaveChangesAsync();
            return ViewModelMapper.ToViewModel(newProjectEntity);
        }

        // delete a project
        public async Task<MessageViewModel> DeleteAsync(int id, bool? deleteNow)
        {
            var project = await FromId(id);

            // If deleteNow is not true, schedule the project for deletion on the first day of the next month.
            if (deleteNow != true)
            {
                // If the project is already scheduled for deletion, throw an exception.
                if (project.DateDeleted.HasValue)
                {
                    throw new OperationNotSupportedException($"Project with id {id} is scheduled to delete on {project.DateDeleted}.");
                }

                // Set the DateDeleted field to the first day of the next month.
                project.DateDeleted = new DateTime(DateTime.Now.Year, DateTime.Now.Month, 1).AddMonths(1);
            }
            else if (project.DateDeleted.HasValue)
            {
                // If deleteNow is true and the project is scheduled for deletion, cancel the deletion.
                project.DateDeleted = null;
            }

            // Remove the project from the Projects DbSet. This will mark the project as deleted in the ChangeTracker.
            _context.Projects.Remove(project);
            await _context.SaveChangesAsync();
            return new MessageViewModel
            {
                Message = $"Project with id {id} is deleted successfully."
            };
        }

        // reset projects table
        public async Task<MessageViewModel> ResetAsync()
        {
            await _context.Projects.ExecuteDeleteAsync();
            return new MessageViewModel
            {
                Message = "Projects table reset successfully."
            };
        }

        private async Task<Project> FromId(int id)
        {
            var projectDb = await _context.Projects.FirstAsync(project => (project.DateDeleted == null || project.DateDeleted >= DateTime.Now) && project.Id == id);
            return projectDb ?? throw new RecordNotFoundException($"Could not find any Project with id: {id}");
        }
    }
}

File: backend-python/app/services/project_service.py
from datetime import date, datetime, timezone

from sqlalchemy import delete, select
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.exceptions import DuplicateEntityException, OperationNotSupportedException, RecordNotFoundException
from app.models.enums import Department, Region
from app.models.project import Project
from app.schemas.common import MessageResponse
from app.schemas.project import ProjectCreate, ProjectResponse, ProjectUpdate
from app.utils.date_utils import first_day_of_next_month
from app.utils.mapper import project_to_response


class ProjectService:
    def __init__(self, db: AsyncSession) -> None:
        self._db = db

    async def get_all(self, department: Department | None, region: Region | None) -> list[ProjectResponse]:
        stmt = select(Project).where(
            (department is None) | ((Project.department.op("&")(int(department))) > 0) if department else True,
            (region is None) | ((Project.region.op("&")(int(region))) > 0) if region else True,
            (Project.date_deleted.is_(None)) | (Project.date_deleted >= datetime.now(timezone.utc)),
        )
        projects = (await self._db.execute(stmt)).scalars().all()
        return [project_to_response(p) for p in projects]

    async def get_all_for_period(
        self,
        year: int,
        month: int,
        department: Department | None,
        region: Region | None,
    ) -> list[ProjectResponse]:
        next_month = first_day_of_next_month(year, month)
        stmt = select(Project).where(
            (department is None) | ((Project.department.op("&")(int(department))) > 0) if department else True,
            (region is None) | ((Project.region.op("&")(int(region))) > 0) if region else True,
            (Project.date_deleted.is_(None)) | (Project.date_deleted >= next_month),
            Project.date_created < next_month,
        )
        projects = (await self._db.execute(stmt)).scalars().all()
        return [project_to_response(p) for p in projects]

    async def get_by_id(self, project_id: int) -> ProjectResponse:
        project = await self._from_id(project_id)
        return project_to_response(project)

    async def create(self, data: ProjectCreate) -> ProjectResponse:
        stmt = select(Project).where(
            Project.date_deleted.is_(None),
            (Project.number == data.number) | (Project.title == data.title),
        )
        existing = (await self._db.execute(stmt)).scalar_one_or_none()
        if existing:
            raise DuplicateEntityException(f"Project already exists in {data.department} department.")

        project = Project(
            number=data.number,
            title=data.title,
            business=data.business,
            department=int(data.department),
            region=int(data.region),
            description=data.description or "",
            date_created=date.today(),
        )
        self._db.add(project)
        await self._db.flush()
        await self._db.refresh(project)
        return project_to_response(project)

    async def update(self, project_id: int, data: ProjectUpdate) -> ProjectResponse:
        project = await self._from_id(project_id)
        if project.date_deleted is not None:
            raise OperationNotSupportedException(
                f"Project with id {project_id} is scheduled to delete on {project.date_deleted}."
            )

        # If number or title change, soft-delete current and create new (mirrors C# logic)
        if data.number is not None or data.title is not None:
            today = date.today()
            if today.month == 1:
                last_prev = date(today.year - 1, 12, 31)
            else:
                from calendar import monthrange
                last_day = monthrange(today.year, today.month - 1)[1]
                last_prev = date(today.year, today.month - 1, last_day)

            project.date_modified = today
            project.date_deleted = datetime.combine(last_prev, datetime.min.time())

            new_project = Project(
                number=data.number or project.number,
                title=data.title or project.title,
                business=data.business or project.business,
                department=int(data.department) if data.department else project.department,
                region=int(data.region) if data.region else project.region,
                description=data.description if data.description is not None else project.description,
                date_created=today,
            )
            self._db.add(new_project)
            await self._db.flush()
            await self._db.refresh(new_project)
            return project_to_response(new_project)

        if data.department is not None:
            project.department = int(data.department)
        if data.description is not None:
            project.description = data.description
        if data.business is not None:
            project.business = data.business
        await self._db.flush()
        await self._db.refresh(project)
        return project_to_response(project)

    async def delete(self, project_id: int, delete_now: bool | None) -> MessageResponse:
        project = await self._from_id(project_id)
        if delete_now is not True:
            if project.date_deleted is not None:
                raise OperationNotSupportedException(
                    f"Project with id {project_id} is scheduled to delete on {project.date_deleted}."
                )
            today = date.today()
            project.date_deleted = datetime.combine(first_day_of_next_month(today.year, today.month), datetime.min.time())
        elif project.date_deleted is not None:
            project.date_deleted = None
        await self._db.delete(project)
        await self._db.flush()
        return MessageResponse(message=f"Project with id {project_id} is deleted successfully.")

    async def reset(self) -> MessageResponse:
        await self._db.execute(delete(Project))
        await self._db.flush()
        return MessageResponse(message="Projects table reset successfully.")

    async def _from_id(self, project_id: int) -> Project:
        stmt = select(Project).where(
            (Project.date_deleted.is_(None)) | (Project.date_deleted >= datetime.now(timezone.utc)),
            Project.id == project_id,
        )
        project = (await self._db.execute(stmt)).scalar_one_or_none()
        if project is None:
            raise RecordNotFoundException(f"Could not find any Project with id: {project_id}")
        return project
